blocks/bricks/bn.py

```diff
@@ -0,0 +1,345 @@
+                            numpy.cast[theano.config.floatX](self.epsilon))
+        assert stdev.broadcastable[1:] == self.population_stdev.broadcastable
+        add_role(mean, BATCH_NORM_MINIBATCH_ESTIMATE)
+        add_role(stdev, BATCH_NORM_MINIBATCH_ESTIMATE)
+        return mean, stdev
+
+    def _prepare_population_statistics(self):
+        mean = _add_batch_axis(self.population_mean, 'population_offset')
+        stdev = _add_batch_axis(self.population_stdev, 'population_divisor')
+        return mean, stdev
+
+    def _allocate(self):
+        input_dim = ((self.input_dim,)
+                     if not isinstance(self.input_dim, collections.Sequence)
+                     else self.input_dim)
+        broadcastable = (tuple(False for _ in range(len(input_dim)))
+                         if self.broadcastable is None else self.broadcastable)
+        if len(input_dim) != len(broadcastable):
+            raise ValueError("input_dim and broadcastable must be same length")
+        var_dim = tuple(1 if broadcast else dim for dim, broadcast in
+                        equizip(input_dim, broadcastable))
+        broadcastable = broadcastable
+
+        # "gamma", from the Ioffe & Szegedy manuscript.
+        self._W = shared_floatx_nans(var_dim, name='batch_norm_scale',
+                                     broadcastable=broadcastable)
+
+        # "beta", from the Ioffe & Szegedy manuscript.
+        self._b = shared_floatx_nans(var_dim, name='batch_norm_shift',
+                                     broadcastable=broadcastable)
+        add_role(self.W, WEIGHT)
+        add_role(self.b, BIAS)
+        self.parameters.append(self.W)
+        self.parameters.append(self.b)
+
+        # These aren't technically parameters, in that they should not be
+        # learned using the same cost function as other model parameters.
+        self.population_mean = shared_floatx_zeros(var_dim,
+                                                   name='population_mean',
+                                                   broadcastable=broadcastable)
+        self.population_stdev = shared_floatx(numpy.ones(var_dim),
+                                              name='population_stdev',
+                                              broadcastable=broadcastable)
+        add_role(self.population_mean, BATCH_NORM_POPULATION_MEAN)
+        add_role(self.population_stdev, BATCH_NORM_POPULATION_STDEV)
+
+    @property
+    def W(self):
+        return self._W
+
+    @property
+    def b(self):
+        return self._b
+
+    def _initialize(self):
+        self.biases_init.initialize(self.b, self.rng)
+        self.weights_init.initialize(self.W, self.rng)
+
+    # Needed for the Feedforward interface.
+    @property
+    def output_dim(self):
+        return self.input_dim
+
+    # The following properties allow for BatchNormalization bricks
+    # to be used directly inside of a ConvolutionalSequence.
```

Comment:
Sure, but my thinking for SpatialBatchNormalization was "a shortcut for doing Batch Normalization *across space*", whereas BatchNormalization normalizes per feature on every axis, and hey, if you want to shove that in a ConvolutionalSequence, that's cool too. (I'd kind of eventually like to see that logic be factored out or else obviated by some sort of change like [this](https://github.com/dwf/blocks/commit/3522f51c2491d5c56cc7345443e87aa24ae959b6), probably not relying on Activation isinstance though).

```python
x = ConvolutionalSequence([..., SpatialBatchNormalization(), ..., SpatialBatchNormalization(), ..., BatchNormalization()])
```
versus
```
x = ConvolutionalSequence([..., SpatialBatchNormalization(), ..., SpatialBatchNormalization(), ..., SpatialBatchNormalization(broadcastable=(False,) * 3)])
```
I think the former wins the cognitive load.